db/map.go

```diff
@@ -0,0 +1,319 @@
+}
+
+func (tx WrappedTransaction) Commit() error {
+	return tx.Transaction.Commit()
+}
+
+func (tx WrappedTransaction) Rollback() error {
+	return tx.Transaction.Rollback()
+}
+
+func (tx WrappedTransaction) Get(holder interface{}, keys ...interface{}) (interface{}, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Get(holder, keys...)
+}
+
+func (tx WrappedTransaction) Insert(list ...interface{}) error {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Insert(list...)
+}
+
+func (tx WrappedTransaction) Update(list ...interface{}) (int64, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Update(list...)
+}
+
+func (tx WrappedTransaction) Delete(list ...interface{}) (int64, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Delete(list...)
+}
+
+func (tx WrappedTransaction) Select(holder interface{}, query string, args ...interface{}) ([]interface{}, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Select(holder, query, args...)
+}
+
+func (tx WrappedTransaction) SelectOne(holder interface{}, query string, args ...interface{}) error {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).SelectOne(holder, query, args...)
+}
+
+func (tx WrappedTransaction) Exec(query string, args ...interface{}) (sql.Result, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Exec(query, args...)
+}
+
+var (
+	// selectTableRegexp matches the table name from an SQL select statement
+	selectTableRegexp = regexp.MustCompile(`(?i)^\s*select\s+[a-z\d\.\(\), \_\*` + "`" + `]+\s+from\s+([a-z\d\_,` + "`" + `]+)`)
+	// insertTableRegexp matches the table name from an SQL insert statement
+	insertTableRegexp = regexp.MustCompile(`(?i)^\s*insert\s+into\s+([a-z\d \_,` + "`" + `]+)\s+(?:set|\()`)
+	// updateTableRegexp matches the table name from an SQL update statement
+	updateTableRegexp = regexp.MustCompile(`(?i)^\s*update\s+([a-z\d \_,` + "`" + `]+)\s+set`)
+	// deleteTableRegexp matches the table name from an SQL delete statement
+	deleteTableRegexp = regexp.MustCompile(`(?i)^\s*delete\s+from\s+([a-z\d \_,` + "`" + `]+)\s+where`)
+
+	// tableRegexps is a list of regexps that tableFromQuery will try to use in
+	// succession to find the table name for an SQL query.
+	tableRegexps = []*regexp.Regexp{
+		selectTableRegexp,
+		insertTableRegexp,
+		updateTableRegexp,
+		deleteTableRegexp,
+	}
+)
+
+// tableFromQuery uses the tableRegexps on the provided query to return the
+// associated table name or an empty string if it can't be determined from the
+// query.
+func tableFromQuery(query string) string {
+	for _, r := range tableRegexps {
+		if matches := r.FindStringSubmatch(query); len(matches) >= 2 {
+			return matches[1]
+		}
+	}
+	return ""
+}
+
+// WrappedExecutor wraps a gorp.SqlExecutor such that its major functions
+// wrap error results in ErrDatabaseOp instances before returning them to the
+// caller.
+type WrappedExecutor struct {
+	gorp.SqlExecutor
```

Comment:
This type definition embeds `gorp.SqlExecutor`, but all the wrapped functions explicitly reference `we.SqlExecutor`. So we have the option of making it a field rather than an embed. That might make it easier in turn to embed WrappedExecutor in WrappedTransaction and WrappedMap, since it would only offer the specific methods that we know we care about.

Or are there other methods that WrappedExecutor does not implement, that we want to pass through to the `gorp.SqlExecutor`?

Reply:
@jsha Changing to a field means `WrappedExecutor` no longer meets `gorp.SqlExecutor` which means it can't be returned from the `WithContext` functions of a transaction or a map, both of which need to return a `gorp.SqlExecutor`. If I change to using a field then `WrappedExecutor` needs to wrap `Query`, `SelectInt`, `SelectNullInt`, etc etc

Reply:
Makese sense.